README.md
# Multi-Tag Picker

Flutter library that combines the text field from `flutter_typeahead` with a row of chips, allowing the user to select from a list of suggestions, as well as creating new values. Similar to `choose_input_chips` except that this widget provides the ability to create new values.

## Usage

### Install

Edit your package's `pubspec.yaml` like so and then run `flutter pub get`:

```
dependencies:
  multi_tag_picker: ^1.0.0
```

### Import

```dart
import 'package:multi_tag_picker/multi_tag_picker.dart';
```

### Example

See the code in `example/lib/main.dart` for a full-fledged working example.

## Similar Projects

The libraries shown below offer form input fields that have something to do with (input) chips. They may be quite different from this library, and that is kind of the point. This library is not meant to be an end-all-be-all to your (input) chip needs, so one of these may offer what you're looking for.

* [awesome_select](https://pub.dev/packages/awesome_select): offers many types of form inputs.
* [chips_choice](https://pub.dev/packages/chips_choice): provides selection of one or more chips.
* [flutter_input_chips](https://pub.dev/packages/flutter_input_chips): text input with free-form creation of new chips.
* [choose_input_chips](https://pub.dev/packages/choose_input_chips): similar to this but the chips are aligned horizontally next to a `Text` and a cursor that simulates editing.
* [simple_chips_input](https://pub.dev/packages/simple_chips_input): text input field with free-form creation of new chips, with optional input validation.

## How to Contribute

Pull requests are welcome. In order to expedite your contributions, please keep the following suggestions in mind. Thank you.

1. Submit small changes one at a time.
1. Keep bug fixes separate from unrelated changes.
1. If you want to reformat the code, do so in a separate commit.
1. If you want to make a lot of changes to bring the code up to date with the latest Dart or Flutter features (such as nullability), do so in a separate commit.
1. Use these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) when writing git commit messages. Bonus points for following the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) convention.

## History

Started as [sarbagyastha/flutter_tagging/](https://github.com/sarbagyastha/flutter_tagging/) and later forked to [culjo/flutter_tagging](https://github.com/culjo/flutter_tagging) and then forked again to [bradintheusa/flutter_tagging_plus](https://github.com/bradintheusa/flutter_tagging_plus) before coming to live here. Wanted to have a source repository that allowed for filing issues.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).
This file follows the convention described at
[Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [1.0.0] - 2023-03-30
**BREAKING CHANGES**
### Renamed from `flutter_tagging_plus` to `multi_tag_picker`.
- Resetting the version number to `1.0.0` to reflect name change.
### Added
* Added optional `placeholderItem` to reserve space for row of chips.
### Changed
* `onChanged()` now receives an immutable copy of the list of values.
* `initialItems` is no longer modified in-place by the widget state.
* Focus stays in text field after selecting or creating a value.
* Added a `SizedBox` of `interiorSpacing` height between text field and chips.
### Removed
* The `Taggable` abstract class and `TaggableMixin`; they do not belong in this library.

## [4.0.1] - 2022-04-20
* Flutter 3.0 fixes.

## [4.0.0] - 2022-04-20
* BradInTheUSA: forked from https://github.com/culjo/flutter_tagging
* culjo: forked from https://github.com/sarbagyastha/flutter_tagging

## [3.0.0] - 2021-03-27
* Null safety release

## [2.2.0+3] - 2020-02-23
* Minor docs changes

## [2.2.0+2] - 2019-12-02
* Changed defaults for `spacing` and `runSpacing` to 8.0

## [2.2.0+1] - 2019-11-29
**BREAKING CHANGES**
* Added mandatory `initialItems` property.

## [2.1.0+2] - 2019-11-28
* Added support for Flutter web platform.

## [2.1.0] - 2019-11-26
**BREAKING CHANGES**
### Changed
* Some properties are renamed and added.
### Added
* Added API documentation.

## [2.0.0+2] - 2019-11-19
* Setting `additionCallback` to `null` disables the feature to add tags.

## [2.0.0+1] - 2019-11-17
**BREAKING CHANGES**
* Complete refactor and improvements.

## [1.1.0] - 2019-03-23
* Updated dependency.
* Added access to newer properties from `flutter_typeahead`.

## [1.0.1] - 2018-12-30
* Added description.

## [1.0.0] - 2018-12-30
* Initial Release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter_tagging_demo/main.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());
  });
}


lib/src/tagging.dart
//
// Copyright 2020 Sarbagya Dhaubanjar
//
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_typeahead/flutter_typeahead.dart';

import 'configurations.dart';

/// A widget that combines a text field provided by the `flutter_typeahead`
/// library and a row of chips that represent the values chosen by the user.
///
/// The type `T` should override the `==` operator as well as `hashCode` to
/// provide suitable equality checks for the values, as they will be stored in a
/// `Set` collection. The `equatable` library is helpful for this purpose. See
/// the `Language` class in `example/lib/main.dart` for an example of
/// implementing this functions.
class FlutterTagging<T> extends StatefulWidget {
  /// Invoked any time values are added or removed.
  final ValueChanged<List<T>> onChanged;

  /// The configuration of the [TextField] that the [FlutterTagging] widget
  /// displays.
  final TextFieldConfiguration textFieldConfiguration;

  /// Called with the search pattern to get the search suggestions.
  ///
  /// This callback is be invoked indirectly by the `flutter_typeahead` widget.
  /// It should return a [List] of suggestions either synchronously, or
  /// asynchronously (as the result of a [Future]). Typically, the list of
  /// suggestions should not contain more than 4 or 5 entries. These entries
  /// will then be provided to [itemBuilder] to display the suggestions.
  final FutureOr<List<T>> Function(String) findSuggestions;

  /// The configuration of [Chip]s that are displayed for selected tags.
  final ChipConfiguration Function(T) configureChip;

  /// The configuration of suggestions displayed when [findSuggestions]
  /// finishes, used to populate the properties of a `ListTile` widget.
  final SuggestionConfiguration Function(T) configureSuggestion;

  /// The configuration of selected tags like their spacing, direction, etc.
  final WrapConfiguration wrapConfiguration;

  /// Defines a new value given the provided query string.
  ///
  /// If null, the tag addition feature is disabled.
  final T Function(String)? additionCallback;

  /// Called when a new tag is being added by the user.
  ///
  /// Optionally may return a refined version of the value to be added to the
  /// list of values.
  final FutureOr<T> Function(T)? onAdded;

  /// Called when waiting for [findSuggestions] to return.
  final Widget Function(BuildContext)? loadingBuilder;

  /// Called when [findSuggestions] returns an empty list.
  final Widget Function(BuildContext)? emptyBuilder;

  /// Called when [findSuggestions] throws an exception.
  final Widget Function(BuildContext, Object?)? errorBuilder;

  /// Called to display animations when [findSuggestions] returns suggestions.
  ///
  /// It is provided with the suggestions box instance and the animation
  /// controller, and expected to return an animation that uses the controller
  /// to display the suggestion box.
  final Widget Function(BuildContext, Widget, AnimationController?)?
      transitionBuilder;

  /// The configuration of suggestion box.
  final SuggestionsBoxConfiguration suggestionsBoxConfiguration;

  /// Spacing between the text field and the row of chips.
  final double interiorSpacing;

  /// The duration that [transitionBuilder] animation takes.
  ///
  /// This argument is best used with [transitionBuilder] and [animationStart]
  /// to fully control the animation.
  ///
  /// Defaults to 500 milliseconds.
  final Duration animationDuration;

  /// The value at which the [transitionBuilder] animation starts.
  ///
  /// This argument is best used with [transitionBuilder] and [animationDuration]
  /// to fully control the animation.
  ///
  /// Defaults to 0.25.
  final double animationStart;

  /// If set to true, no loading box will be shown while suggestions are
  /// being fetched. The [loadingBuilder] will also be ignored.
  ///
  /// Defaults to false.
  final bool hideOnLoading;

  /// If set to true, nothing will be shown if there are no results. The
  /// [emptyBuilder] will also be ignored.
  ///
  /// Defaults to false.
  final bool hideOnEmpty;

  /// If set to true, nothing will be shown if there is an error. The
  /// [errorBuilder] will also be ignored.
  ///
  /// Defaults to false.
  final bool hideOnError;

  /// If defined, an invisible placeholder based on this item will occupy the
  /// space where the row of chips would be when there are no values selected.
  final T? placeholderItem;

  /// The duration to wait after the user stops typing before calling
  /// [findSuggestions].
  ///
  /// This duration is set by default to 300 milliseconds.
  final Duration debounceDuration;

  /// If set to true, suggestions will be fetched immediately when the field is
  /// added to the view.
  ///
  /// The suggestions box will only be shown when the field receives focus. To
  /// make the field receive focus immediately, you can set the `autofocus`
  /// property in the [textFieldConfiguration] to true.
  ///
  /// Defaults to false.
  final bool enableImmediateSuggestion;

  /// List of inital values, if any.
  final List<T> initialItems;

  /// Creates a [FlutterTagging] widget.
  const FlutterTagging({
    Key? key,
    required this.findSuggestions,
    required this.configureChip,
    required this.configureSuggestion,
    required this.onChanged,
    this.initialItems = const [],
    this.additionCallback,
    this.enableImmediateSuggestion = false,
    this.errorBuilder,
    this.loadingBuilder,
    this.emptyBuilder,
    this.placeholderItem,
    this.wrapConfiguration = const WrapConfiguration(),
    this.textFieldConfiguration = const TextFieldConfiguration(),
    this.suggestionsBoxConfiguration = const SuggestionsBoxConfiguration(),
    this.interiorSpacing = 8.0,
    this.transitionBuilder,
    this.debounceDuration = const Duration(milliseconds: 300),
    this.hideOnEmpty = false,
    this.hideOnError = false,
    this.hideOnLoading = false,
    this.animationDuration = const Duration(milliseconds: 500),
    this.animationStart = 0.25,
    this.onAdded,
  }) : super(key: key);

  @override
  State<FlutterTagging<T>> createState() => _FlutterTaggingState<T>();
}

class _FlutterTaggingState<T> extends State<FlutterTagging<T>> {
  late final TextEditingController _textController;
  late final FocusNode _focusNode;
  final Set<T> _values = <T>{};
  T? _additionItem;

  @override
  void initState() {
    super.initState();
    _values.addAll(widget.initialItems);
    _textController =
        widget.textFieldConfiguration.controller ?? TextEditingController();
    _focusNode = widget.textFieldConfiguration.focusNode ?? FocusNode();
  }

  @override
  void dispose() {
    _textController.dispose();
    _focusNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisAlignment: MainAxisAlignment.start,
      children: [
        TypeAheadField<T>(
          getImmediateSuggestions: widget.enableImmediateSuggestion,
          debounceDuration: widget.debounceDuration,
          hideOnEmpty: widget.hideOnEmpty,
          hideOnError: widget.hideOnError,
          hideOnLoading: widget.hideOnLoading,
          animationStart: widget.animationStart,
          animationDuration: widget.animationDuration,
          autoFlipDirection:
              widget.suggestionsBoxConfiguration.autoFlipDirection,
          direction: widget.suggestionsBoxConfiguration.direction,
          hideSuggestionsOnKeyboardHide:
              widget.suggestionsBoxConfiguration.hideSuggestionsOnKeyboardHide,
          keepSuggestionsOnLoading:
              widget.suggestionsBoxConfiguration.keepSuggestionsOnLoading,
          keepSuggestionsOnSuggestionSelected: widget
              .suggestionsBoxConfiguration.keepSuggestionsOnSuggestionSelected,
          suggestionsBoxController:
              widget.suggestionsBoxConfiguration.suggestionsBoxController,
          suggestionsBoxDecoration:
              widget.suggestionsBoxConfiguration.suggestionsBoxDecoration,
          suggestionsBoxVerticalOffset:
              widget.suggestionsBoxConfiguration.suggestionsBoxVerticalOffset,
          errorBuilder: widget.errorBuilder,
          transitionBuilder: widget.transitionBuilder,
          loadingBuilder: (context) =>
              widget.loadingBuilder?.call(context) ??
              const SizedBox(
                height: 3.0,
                child: LinearProgressIndicator(),
              ),
          noItemsFoundBuilder: widget.emptyBuilder,
          textFieldConfiguration: widget.textFieldConfiguration.copyWith(
            focusNode: _focusNode,
            controller: _textController,
            enabled: widget.textFieldConfiguration.enabled,
          ),
          suggestionsCallback: (query) async {
            var suggestions = await widget.findSuggestions(query);
            suggestions =
                suggestions.where((r) => !_values.contains(r)).toList();
            if (widget.additionCallback != null && query.isNotEmpty) {
              final additionItem = widget.additionCallback!(query);
              if (!suggestions.contains(additionItem) &&
                  !_values.contains(additionItem)) {
                _additionItem = additionItem;
                suggestions.insert(0, additionItem);
              } else {
                _additionItem = null;
              }
            }
            return suggestions;
          },
          itemBuilder: (context, item) {
            final conf = widget.configureSuggestion(item);
            return ListTile(
              key: ObjectKey(item),
              title: conf.title,
              subtitle: conf.subtitle,
              leading: conf.leading,
              trailing: InkWell(
                splashColor: conf.splashColor ?? Theme.of(context).splashColor,
                borderRadius: conf.splashRadius,
                onTap: () async {
                  final itemItem = widget.onAdded != null
                      ? await widget.onAdded!(item)
                      : item;
                  setState(() {
                    _values.add(itemItem);
                  });
                  widget.onChanged.call(_values.toList(growable: false));
                  _textController.clear();
                },
                child: Builder(
                  builder: (context) {
                    if (conf.additionWidget != null && _additionItem == item) {
                      return conf.additionWidget!;
                    } else {
                      return const SizedBox(width: 0);
                    }
                  },
                ),
              ),
            );
          },
          onSuggestionSelected: (suggestion) {
            if (_additionItem != suggestion) {
              setState(() {
                _values.add(suggestion);
              });
              widget.onChanged.call(_values.toList(growable: false));
              _textController.clear();
              // focus change must happen after this task completes
              Future.delayed(Duration.zero, () {
                _focusNode.requestFocus();
              });
            }
          },
        ),
        if (widget.interiorSpacing > 0)
          SizedBox(height: widget.interiorSpacing),
        Wrap(
          alignment: widget.wrapConfiguration.alignment,
          crossAxisAlignment: widget.wrapConfiguration.crossAxisAlignment,
          runAlignment: widget.wrapConfiguration.runAlignment,
          runSpacing: widget.wrapConfiguration.runSpacing,
          spacing: widget.wrapConfiguration.spacing,
          direction: widget.wrapConfiguration.direction,
          textDirection: widget.wrapConfiguration.textDirection,
          verticalDirection: widget.wrapConfiguration.verticalDirection,
          children: _values.isEmpty && widget.placeholderItem != null
              ? _buildPlaceholder(
                  context,
                  // ignore: null_check_on_nullable_type_parameter
                  widget.configureChip(widget.placeholderItem!),
                )
              : _values.map<Widget>((item) {
                  final conf = widget.configureChip(item);
                  return Chip(
                    label: conf.label,
                    shape: conf.shape,
                    avatar: conf.avatar,
                    backgroundColor: conf.backgroundColor,
                    clipBehavior: conf.clipBehavior,
                    deleteButtonTooltipMessage: conf.deleteButtonTooltipMessage,
                    deleteIcon: conf.deleteIcon,
                    deleteIconColor: conf.deleteIconColor,
                    elevation: conf.elevation,
                    labelPadding: conf.labelPadding,
                    labelStyle: conf.labelStyle,
                    materialTapTargetSize: conf.materialTapTargetSize,
                    padding: conf.padding,
                    shadowColor: conf.shadowColor,
                    onDeleted: () {
                      setState(() {
                        _values.remove(item);
                      });
                      widget.onChanged.call(_values.toList(growable: false));
                    },
                  );
                }).toList(),
        ),
      ],
    );
  }
}

List<Widget> _buildPlaceholder(BuildContext context, ChipConfiguration conf) {
  return [
    Visibility(
      visible: false,
      maintainAnimation: true,
      maintainSize: true,
      maintainState: true,
      child: Chip(
        label: conf.label,
        shape: conf.shape,
        avatar: conf.avatar,
        clipBehavior: conf.clipBehavior,
        deleteIcon: conf.deleteIcon,
        elevation: conf.elevation,
        labelPadding: conf.labelPadding,
        labelStyle: conf.labelStyle,
        padding: conf.padding,
      ),
    )
  ];
}


lib/src/taggable.dart
//
// Copyright 2020 Sarbagya Dhaubanjar
//
import 'package:collection/collection.dart';
import 'package:meta/meta.dart';

/// A base class to facilitate [operator==] and [hashCode] overrides.
///
/// ```dart
/// class ConstTest extends Taggable {
///   const ConstTest(this.a);
///
///   final int a;
///
///   @override
///   List<Object> get props => [a];
/// }
/// ```
@immutable
abstract class Taggable {
  /// The [List] of `props` (properties) which will be used to determine whether
  /// two [Taggables] are equal.
  List<Object> get props;

  /// If true, string comparison will be case sensitive.
  bool get caseSensitive => false;

  /// A class that helps implement equality
  /// without needing to explicitly override == and [hashCode].
  /// Taggables override their own `==` operator and [hashCode] based on their `props`.
  const Taggable();

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Taggable &&
          runtimeType == other.runtimeType &&
          _equals(props, other.props, caseSensitive);

  @override
  int get hashCode => runtimeType.hashCode ^ _mapPropsToHashCode(props);

  @override
  String toString() => '$runtimeType';
}

/// You must define the [TaggableMixin] on the class
/// which you want to make Taggable.
///
/// [TaggableMixin] does the override of the `==` operator as well as `hashCode`.
mixin TaggableMixin implements Taggable {
  @override
  List<Object> get props;

  @override
  bool get caseSensitive => false;

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        other is TaggableMixin &&
            runtimeType == other.runtimeType &&
            _equals(props, other.props, caseSensitive);
  }

  @override
  int get hashCode => runtimeType.hashCode ^ _mapPropsToHashCode(props);

  @override
  String toString() => '$runtimeType';
}

int _mapPropsToHashCode(dynamic props) {
  var hashCode = 0;

  if (props is Map) {
    props.forEach((key, value) {
      final propHashCode =
          _mapPropsToHashCode(key) ^ _mapPropsToHashCode(value);
      hashCode = hashCode ^ propHashCode;
    });
  } else if (props is List || props is Iterable || props is Set) {
    props.forEach((prop) {
      final propHashCode =
          (prop is List || prop is Iterable || prop is Set || prop is Map)
              ? _mapPropsToHashCode(prop)
              : prop.hashCode;
      hashCode = hashCode ^ propHashCode;
    });
  } else {
    hashCode = hashCode ^ props.hashCode;
  }

  return hashCode;
}

const DeepCollectionEquality _equality = DeepCollectionEquality();

bool _equals(List list1, List list2, bool caseSensitive) {
  if (identical(list1, list2)) return true;
  final length = list1.length;
  if (length != list2.length) return false;

  for (var i = 0; i < length; i++) {
    final unit1 = list1[i];
    final unit2 = list2[i];

    if (unit1 is Iterable || unit1 is List || unit1 is Map || unit1 is Set) {
      if (!_equality.equals(unit1, unit2)) return false;
    } else if (unit1?.runtimeType != unit2?.runtimeType) {
      return false;
    } else if (unit1 is String) {
      if (caseSensitive && unit1.compareTo(unit2) != 0) {
        return false;
      }
      if (!caseSensitive &&
          unit1.toLowerCase().compareTo(unit2.toLowerCase()) != 0) {
        return false;
      }
    } else if (unit1 != unit2) {
      return false;
    }
  }
  return true;
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_syntax_view/flutter_syntax_view.dart';
import 'package:multi_tag_picker/multi_tag_picker.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'MultiPicker Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _selectedValuesJson = 'are yet to be found...';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('MultiPicker Demo'),
      ),
      body: Column(
        children: <Widget>[
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: FlutterTagging<Language>(
              textFieldConfiguration: const TextFieldConfiguration(
                decoration: InputDecoration(
                  border: InputBorder.none,
                  filled: true,
                  hintText: 'Enter a query',
                  labelText: 'Select Tags',
                ),
              ),
              findSuggestions: getLanguages,
              additionCallback: (value) {
                return Language(
                  name: value,
                  position: 0,
                );
              },
              configureSuggestion: (lang) {
                return SuggestionConfiguration(
                  title: Text(lang.name),
                  subtitle: Text(lang.position.toString()),
                  additionWidget: const Chip(
                    avatar: Icon(
                      Icons.add_circle,
                    ),
                    label: Text('Add New Tag'),
                  ),
                );
              },
              placeholderItem: const Language(
                name: 'dummy',
                position: 0,
              ),
              configureChip: (lang) {
                return ChipConfiguration(
                  label: Text(lang.name),
                );
              },
              onChanged: (values) {
                setState(() {
                  if (values.isEmpty) {
                    _selectedValuesJson = 'have left us...';
                  } else {
                    _selectedValuesJson = values
                        .map<String>((lang) => '\n${lang.toJson()}')
                        .toList()
                        .toString()
                        .replaceFirst('}]', '}\n]');
                  }
                });
              },
            ),
          ),
          const Text(
            'The Chosen',
            style: TextStyle(fontSize: 18.0),
          ),
          Expanded(
            child: Padding(
              padding: const EdgeInsets.all(8.0),
              child: Container(
                decoration: BoxDecoration(
                  border: Border.all(width: 1.0),
                ),
                child: SyntaxView(
                  code: _selectedValuesJson,
                  syntax: Syntax.JAVASCRIPT,
                  fontSize: 16.0,
                  expanded: true,
                  withLinesCount: false,
                  syntaxTheme: SyntaxTheme.standard(),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

/// Mocks fetching language from network API with delay of 500ms.
Future<List<Language>> getLanguages(String query) async {
  if (query.isEmpty) {
    return const [];
  }
  await Future.delayed(const Duration(milliseconds: 500), null);
  final lowercaseQuery = query.toLowerCase();
  return const <Language>[
    Language(name: 'Rust', position: 1),
    Language(name: 'Dart', position: 2),
    Language(name: 'Scheme', position: 3),
    Language(name: 'Swift', position: 4),
    Language(name: 'Erlang', position: 5),
    Language(name: 'Go', position: 6),
    Language(name: 'F#', position: 7),
    Language(name: 'Java', position: 8),
    Language(name: 'C#', position: 9),
    Language(name: 'JavaScript', position: 10),
    Language(name: 'APL', position: 11),
    Language(name: 'Python', position: 12),
    Language(name: 'Haxe', position: 13),
    Language(name: 'Io', position: 14),
    Language(name: 'Lua', position: 15),
    Language(name: 'OCaml', position: 16),
    Language(name: 'C++', position: 17),
    Language(name: 'Prolog', position: 18),
    Language(name: 'TypeScript', position: 19),
    Language(name: 'Zig', position: 20),
  ]
      .where((lang) => lang.name.toLowerCase().contains(lowercaseQuery))
      .toList(growable: false)
    ..sort((a, b) => a.name
        .toLowerCase()
        .indexOf(lowercaseQuery)
        .compareTo(b.name.toLowerCase().indexOf(lowercaseQuery)));
}

/// Language represents a programming language.
///
/// Note that this class overrides the `==` operator and `hashCode` method in
/// order to provide sensible equality with other values of the same type.
class Language {
  final String name;
  final int position;

  const Language({
    required this.name,
    required this.position,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Language &&
          runtimeType == other.runtimeType &&
          name == other.name;

  @override
  int get hashCode => name.hashCode;

  @override
  String toString() {
    return name;
  }

  String toJson() => '''  {
    "name": $name,\n
    "position": $position\n
  }''';
}
